### PR TITLE
Support multithreading in `seq_join`/`parallel_join`

### PR DESCRIPTION
**Pull request by akoshelev**
Support is currently behind a feature flag that is not enabled by default We use userspace concurrency to drive many futures in parallel by spawning tasks inside the executor. This model is not ideal for performance because memory loads will happen across thread boundaries and NUMA cores, but already gives 50% more throughput for OPRF version and 200% to old IPA.

This is just a draft for folks to take a look and provide feedback. A few things are not working and need to be fixed before pushing this to main

* `async_scoped` may resolve futures out of order. Under the hood, it uses `FuturesUnordered` and streams elements out of it when collected. This is not ideal for our `seq_join` where we expect things to be in FIFO order. I am planning to raise a PR and suggest to use `FuturesOrdered`. If that does not go through, a bit of work is required on our side to annotate futures with index and keeping a buffer.
* `shuttle` flags our IPA tests saying it exceeded the limit of 100k steps. It is a bit suspicious as it may (or may not) indicate a cycle. I need to look deeper into it.

### CODE REVIEW

File: ipa-core/src/seq_join.rs [hunk 1]
```diff
@@ -95,209 +113,348 @@ pub trait SeqJoin {
...
+        F: Future<Output = Result<O, E>> + Send + 'fut,
+        O: Send + 'static,
+        E: Send + 'static,
+    {
+        // TODO: implement spawner for shuttle
+        let mut scope = {
+            let iter = iterable.into_iter();
+            // SAFETY: scope object does not escape this function. All futures are driven to
+            // completion inside it or cancelled if a panic occurs.
+            let mut scope = unsafe { create_spawner() };
```
Comment (andyleiserson) [inline ipa-core/src/seq_join.rs:420]: The async stuff makes it harder to reason about "escapes this function". If you view the function as being active until the boxed future completes, then yes, I think that's true. But if anyone calls `mem::forget` for the boxed future before it completes, then there could be a memory safety violation.

It is my understanding though, that as long as you don't `mem::forget` the future (and thus instead always either drive it to completion or drop it), then there can't be a memory safety violation. Maybe it would make sense to put `mem::forget` on the disallowed function list like `join_all`?

I also thought about whether we could use the safe `spawn_and_block` API. That would mean consuming a thread per active query, which seems tolerable. But if we want to avoid calling `spawn_and_block` from a regular tokio executor, and instead call it via `thread::spawn` or `tokio::task::spawn_blocking`, that brings us right back to the `'static` problem. The only way I can see it working is if we spawn some number of tasks that receive futures over channels in `query::executor::do_query` (which is where the query lifetime is realized -- it is already calling `tokio::spawn`).

Comment (andyleiserson) [reply ipa-core/src/seq_join.rs:420]: Or maybe `thread::scope` helps us here?

File: ipa-core/src/seq_join.rs [hunk 2]
```diff
@@ -667,4 +667,50 @@ mod test {
...
+
+        let iterable = (0..N)
+            .map(|i| {
+                let borrowed = &borrow_from_me;
+                let b1 = barrier1.clone();
+                let b2 = barrier2.clone();
+                async move {
+                    b1.wait().await;
+                    for _ in 0..100 {
+                        if borrowed != &vec![1, 2, 3] {
```
Comment (martinthomson) [inline ipa-core/src/seq_join.rs:694]: Can you deal with this differently, for instance, by passing in a weak reference that should remain live while the outside is operating.  That would be less prone to problems of scope.

File: ipa-core/src/seq_join.rs (hunk 1 above)
Comment (akoshelev) [reply ipa-core/src/seq_join.rs]: You are right - I mistakenly assumed that not leaking futures created within `Scope` makes it safe. It does not and I wrote a test to confirm that. 

> It is my understanding though, that as long as you don't mem::forget the future (and thus instead always either drive it to completion or drop it), then there can't be a memory safety violation. Maybe it would make sense to put mem::forget on the disallowed function list like join_all?

that's my plan - restricting `mem::forget` won't be enough, we also need to monitor usages of `Box::leak` and `ManuallyDrop`. I wouldn't go down that path if we were a library, but for IPA as an application this could be acceptable mitigation along with sanitizers run. 

> I also thought about whether we could use the safe spawn_and_block API. 

I am going to play more with it but I don't feel that we can make it work. We combine the use of `seq_join` with `parallel_join` inside it and blocking the thread may lead to some weird deadlocks. My current thinking that it is probably not worth trying to make it safe, given that `seq_join` is inherently unsafe and cannot be make safe.

**Comment by akoshelev** [comment]
It is ready for review now. I added Miri tests and currently importing `async_scoped` from github because 0.9 hasn't been released yet. 

@martinthomson @andyleiserson please take a look if you have time

**Review by martinthomson — APPROVED** [review]
This is pretty good stuff.

However, I have some reservations about the ultimate performance profile of this. That's not enough to say not to land this, but it's enough to suggest that maybe there is some new work in our futures (no pun intended).

If I understand how async-scoped works, spawning tasks will put them into a pool somewhere.  Everything in that pool is polled, according to whatever scheduler logic applies.

It seems like the stream on the `TokioScope` ensures that spawned tasks resolve in the order in which they were added, which makes our `parallel_join` logic easier.  That's the change you managed to arrange for async-scoped, which is good.

I'm interested in whether we need that sort of polling logic for a sequential join or whether something a whole lot simpler might also work.

For a sequential join, the core problem we have is that the source stream is not polled reliably.  For that, we need just a single task to be spawned, one that drives the stream to completion.

The trick there is that there is no way to get the items from that stream back to the current task for polling.  But decoupling could use a single `mpsc::channel` to allow that to happen (I want single producer, single consumer, but apparently that's not a thing).

Basically, you would create a channel, pass the sender to a spawned task that reads from the source stream and sends the results.  Then the current task would read from the receiver as we do in the single-threaded (and therefore deadlock-y) variant.

The benefit of that sort of approach is that you would be able to keep the source stream fed with polling from the spawned task, but you don't need to engage most of the nasty async-scoped stuff.

We might still need the crate so that we can hide the lifetime and panic horrors that are associated; after all, we would be no less exposed to the drop issues.

I'm trying to remember now why I previously thought that maybe that sort of approach wasn't wise, but looking into it now, I can't see why.

The primary reason I suggest this is that while the overhead from `FuturesOrdered` is probably minimal, we're going to be loading those up pretty hard and maybe we can avoid needing multiple collection of futures that hold significant numbers of entries.  The channel approach also greatly reduces the number of futures on hand overall, which might be a good trade.

File: ipa-core/src/lib.rs [hunk 3]
```diff
@@ -118,7 +118,10 @@ pub(crate) mod test_executor {
         Fut: Future<Output = ()>,
     {
         tokio::runtime::Builder::new_multi_thread()
-            .enable_all()
+            // IO driver is disabled to run our tests under Miri. If you need it, make sure you
+            // annotate this test with #[cfg(not(miri))]
+            // https://github.com/rust-lang/miri/issues/2057
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply ipa-core/src/lib.rs:123]: yea let me rephrase it

File: .github/workflows/check.yml [hunk 4]
```diff
@@ -156,7 +160,21 @@ jobs:
       - name: Add Rust sources
         run: rustup component add rust-src
       - name: Run tests with sanitizer
-        run: RUSTFLAGS="-Z sanitizer=${{ matrix.sanitizer }} -Z sanitizer-memory-track-origins" cargo test -Z build-std --target $TARGET --no-default-features --features "cli web-app real-world-infra test-fixture descriptive-gate"
+        run: RUSTFLAGS="-Z sanitizer=${{ matrix.sanitizer }} -Z sanitizer-memory-track-origins" cargo test -Z build-std --target $TARGET --no-default-features --features "cli web-app real-world-infra test-fixture descriptive-gate ${{ matrix.features }}"
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply .github/workflows/check.yml:163]: oh no, I am not that brave. We run sanitizers with multi-threading feature enabled [as well](https://github.com/private-attribution/ipa/actions/runs/7615944630/job/20741595617)

File: ipa-core/src/seq_join.rs [hunk 5]
```diff
@@ -365,8 +534,8 @@ mod test {
     }
 
     /// A fully synchronous test with a synthetic stream, all the way to the end.
-    #[test]
-    fn complete_stream() {
+    #[tokio::test]
+    async fn complete_stream() {
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply ipa-core/src/seq_join.rs:537]: I tried to make it work with executor-spawned tasks, but failed and forgot to change it back

File: ipa-core/src/seq_join.rs [hunk 6]
```diff
@@ -95,209 +128,345 @@ pub trait SeqJoin {
...
-{
-    type Item = F::Output;
+        /// Takes the resolved value out
+        ///
+        /// ## Panics
+        /// If the value is not ready yet.
+        #[must_use]
+        fn take(self) -> F::Output {
+            let ActiveItem::Resolved(v) = self else {
+                panic!("No value to take out");
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply ipa-core/src/seq_join.rs:203]: I am more inclined to make this unreachable because it is impossible to trigger this code path by abusing `SequentialFutures` public API.

**Comment by akoshelev** [comment]
> The primary reason I suggest this is that while the overhead from `FuturesOrdered` is probably minimal, we're going to be loading those up pretty hard and maybe we can avoid needing multiple collection of futures that hold significant numbers of entries. The channel approach also greatly reduces the number of futures on hand overall, which might be a good trade.

I agree, we are not doing anything differently here - `FuturesOrdered` is used inside `futures::try_join_all` for the same purpose, so there should be no added cost of managing the FIFO order.

File: ipa-core/src/seq_join.rs [hunk 7]
```diff
@@ -408,3 +577,191 @@ mod test {
...
+            assert_eq!((1..5).collect::<Vec<_>>(), res);
+        });
+    }
+
+    /// This test has to use multi-threaded runtime because early return causes `TryCollect` to be
+    /// dropped and the remaining futures to be cancelled which can only happen if there is more
+    /// than one thread available.
+    ///
+    /// This behavior is only applicable when `seq_try_join_all` uses more than one thread, for
+    /// maintenance reasons, we use it even parallelism is turned off.
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply ipa-core/src/seq_join.rs]: This comment is outdated and no longer relevant

File: ipa-core/src/seq_join.rs [hunk 8]
```diff
@@ -408,3 +577,191 @@ mod test {
...
+            let bad_accesses = Arc::new(tokio::sync::Barrier::new(N + 1));
+
+            let futures = (0..N)
+                .map(|_| {
+                    let borrowed = Arc::downgrade(&borrow_from_me);
+                    let start = start.clone();
+                    let bad_access = bad_accesses.clone();
+                    async move {
+                        start.wait().await;
+                        // at this point, the parent future is forgotten and borrowed should point to nothing
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply ipa-core/src/seq_join.rs]: yep, changed the wording here

**Comment by akoshelev** [comment]
> I'm interested in whether we need that sort of polling logic for a sequential join or whether something a whole lot simpler might also work.

Here are some numbers that steered me towards using more than a single thread for seq_join

```
single thread:

Running IPA for 100000 records took 1245s

multithreaded seq_join with metrics disabled

Running IPA for 100000 records took 345.480439625s
```

So I think we should use Tokio executor for `seq_join`. However, I agree that long term I would like to see parallelism applied somewhere else. We will hit the scaling bottleneck (if we haven't already) soon because of allocation thrashing and random read/write access across many threads. A much better approach imo would be aligning every step with its own task, making sure allocations there stay local all the way to the kernel stack (including RSS/XPS). This would be a much more involved change, but worth considering when we don't have low-hanging fruits anymore

**Review by andyleiserson — APPROVED** [review]

**Comment by andyleiserson** [comment]
Re: @martinthomson's proposal to spawn a task that polls the seq_join input stream and sends to a channel:

The issue I see here is managing/controlling how far forward that task can get. If the channel has unlimited capacity, then it potentially uses as much memory as a non-streaming implementation. If the channel has limited capacity, then the question is how big the channel needs to be to prevent deadlocks -- once the channel is full, the task will stop polling the input stream, same as now. To me it feels pretty similar to the buffering solution.
